### PR TITLE
Build both debug and release libmono-android regardless of config.

### PR DESCRIPTION
**Pull request by atsushieno**
Whether you build xamarin-android in debug mode or release mode
should not matter when user wants to build app either in debug or
release mode. Current build system builds only debug mono-android for
Debug configuration and release mono-android for Release configuration,
which only makes it useless and annoying to diagnose the issues that
depends on app build configuration.

Build both libmono-android regardless of xamarin-android builder's
configuration and make it work.

### CODE REVIEW

**Comment by dellis1972** [comment]
looks ok to me

File: src/monodroid/monodroid.targets [hunk 1]
```diff
@@ -22,22 +19,33 @@
...
-        DestinationFiles="@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.$(_Conf).d.so')"
+        DestinationFiles="@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.debug.d.so')"
     />
     <Copy
         SourceFiles="@(AndroidSupportedTargetJitAbi->'libs\%(Identity)\libmonodroid.so')"
-        DestinationFiles="@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.$(_Conf).so')"
+        DestinationFiles="@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.debug.so')"
+    />
+    <Exec
+        Command="$(AndroidToolchainDirectory)\ndk\ndk-build CONFIGURATION=Release V=1"
```
Comment (jonpryor) [inline src/monodroid/monodroid.targets]: Do we need to do a "clean" or something before re-running `ndk-build`, to ensure that everything is rebuilt? I fear/suspect that simply changing an environment variable won't cause everything to be recompiled.

The "best" fix would be for Debug and Release object files to go in separate directories, but I couldn't easily determine how to control that within `ndk-build`... :-/

Comment (atsushieno) [inline src/monodroid/monodroid.targets]: That's actually a good idea and I'd do that too, but IMO that should be in different commit :-)

Comment (atsushieno) [inline src/monodroid/monodroid.targets]: I mean, ndk-build clean part. Not sure about separate directory yet.

Comment (jonpryor) [inline src/monodroid/monodroid.targets]: That doesn't answer the question: Do we need to "clean" to ensure that things are rebuilt? To rephrase it, what happens when you do:

    $ ndk-build CONFIGURATION=Debug V=1
    $ ndk-build CONFIGURATION=Release V=1

...since that's what this PR *does*.

*For me*, it does...nothing:

    $ ndk-build CONFIGURATION=Debug
    [armeabi-v7a] Install        : libmonodroid.so => libs/armeabi-v7a/libmonodroid.so
    $ ndk-build CONFIGURATION=Release
    [armeabi-v7a] Install        : libmonodroid.so => libs/armeabi-v7a/libmonodroid.so

Changing `CONFIGURATION` does *not* cause `ndk-build` to rebuild *anything* if the outputs are already up-to-date.

Meaning your "Release" *won't actually be a release build!*

(This is *probably* why it even builds for you *at all*.)

To force a rebuild with configuration change, an intermediate clean is required:

    $ ndk-build CONFIGURATION=Debug
    ...
    $ ndk-build clean
    $ ndk-build CONFIGURATION=Release

...which *fails* for me:

    In file included from jni/../../../external/mono/support/nl.h:3:0,
                     from jni/../../../external/mono/support/nl.c:14:
    jni/../../../external/mono/eglib/src/glib.h:22:26: fatal error: eglib-config.h: No such file or directory
     #include <eglib-config.h>
                              ^
    compilation terminated.
    make: *** [obj/local/armeabi-v7a/objs/monodroid/__/__/__/external/mono/support/nl.o] Error 1

Which makes sense: it's trying to use `$(CONFIGURATION)` to grab `eglib-config`, from `$(LOCAL_PATH)/../../../build-tools/mono-runtimes/obj/$(CONFIGURATION)/$(TARGET_ARCH_ABI)/eglib/src`, and if we haven't build Mono in the Release configuration -- which we *haven't* -- then this can't possibly work.

Comment (jonpryor) [inline src/monodroid/monodroid.targets]: What this PR *wants* to do is generate both `libmono-android.debug.so` and `libmono-android.release.so`. As per above, this can't be achieved by using `CONFIGURATION`: it won't build.

Instead, I would suggest adding a separate make variable to `jni/Android.mk` which controls whether `-DRELEASE=1` is used, *separate from* `$(CONFIGURATION)`. This would allow you to do:

    $ ndk-build CONFIGURATION=Debug FASTDEV=1 V=1
    # copy to libmono-android.debug.so
    $ ndk-build clean
    $ ndk-build CONFIGURATION=Debug FASTDEV=0 V=1
    # copy to libmono-android.release.so

Comment (atsushieno) [inline src/monodroid/monodroid.targets]: Note that the current Clean logic is the same as it USED TO BE.

**Comment by jonpryor** [comment]
I believe that I'm not being sufficiently clear, but I'm not sure how to further clarify. :-(

> The reason why whatever you suggested should be done in SEPARATE COMMIT is, changeset atomicity

Atomicity of *what*? *Presumably*, atomicity of a *working feature*, yes? The world "works" in a well-defined manner before the commit, and "works" in a well-defined manner after the commit, with no breakage in behavior? (Alas, this project can't satisfy that, because *nothing* works *now*, so there's no well-defined "pre-commit" state, which we can all agree is extremely frustrating...)

I believe that the purpose of atomicity is that the commit itself should be self-contained, self-consistent, and...*work*.

Which is my problem with this PR: it doesn't work. It doesn't do what it says on the proverbial tin. It is, in fact, *broken*, but broken in such a way that the breakage isn't immediately self evident.

That's not reason to commit a broken "atomic" commit and fix the brokenness in a later "atomic" commit. Fix the problem now, so that it doesn't require two commits.

> Note that the current Clean logic is the same as it USED TO BE.

I'm not talking about the `Clean` target.

**Comment by jonpryor** [comment]
What is the purpose of this PR? It's in the initial description:

> Whether you build xamarin-android in debug mode or release mode should not matter when user wants to build app either in debug or release mode

The fix? We need to build both `libmono-android.debug.so` and `libmono-android.release.so` files.

That's the purpose.

There are two ways to satisfy this purpose:

1. Copy the `debug` variant to the `release` variant.

        ndk-build CONFIGURATION=$(Configuration) V=1
        # copy out obj\local\%(Identity)\libmonodroid.so to libmono-android.debug.so
        # copy out obj\local\%(Identity)\libmonodroid.so to libmono-android.release.so

    This ensures that both `libmono-android.debug.so` and `libmono-android.release.so` exist. That satisfies the purpose of the PR.

    The problem with this is that the "release" variant is "wrong": the `debug` variant supports the `.__override__` directory, the `release` version *shouldn't*, but in this solution the `release` variant *could* support `.__override__`, which is undesirable.

2. Rebuild the project so that the native library is built with `-DRELEASE=1`.

(2) is the "correct" approach. It would ensure that the `release` version doesn't support the `.__override__` directory, *and* ensures that both `libmono-android.debug.so` and `libmono-android.release.so` both exist, which satisfies the stated purpose of the PR.

The problem is that this PR doesn't do (2). It does (1) but *looks like it's trying to do (2)*.

Please, *please*, understand this point: when `ndk-build` is run twice, the second invocation is a no-op. It does *nothing*. Thus, even though you invoke `ndk-build` a second time, before copying and creating the `libmono-android.release.so` files, *you're not actually creating a library built with `-DRELEASE=1`*. The two libraries are *identical*.

The PR description screams out "This should do (2)", the patch *looks like* it does (2), but it's actually doing (1)!

Atomicity-wise, isn't this bad? Shouldn't it *actually* implement (2) when the PR and commit message states that it's doing (2)?

If (1) is *actually* desirable -- and perhaps it is! -- then the commit should say so and look like it. The second `ndk-build` invocation should be *removed*, and the commit message altered to make it explicitly clear that both the `debug` and `release`  variants are in fact *the same file*.

**Comment by atsushieno** [comment]
Oh, I thought you were talking about /t:Clean target. Alright, that build part is indeed problematic. Thanks for iterating the explanation.

With all those in mind, yeah it would make everything easier and consistent to have separate builds for Debug and Release. I'll find the way to achieve that.

**Comment by atsushieno** [comment]
I'm going to create another PR for this.
